File: src/main/java/org/app/infrastructure/utils/Pageable.java
package org.app.infrastructure.utils;

import org.app.domain.exceptions.PageSizeTooLargeException;

public class Pageable {
    private final int offset;
    private final int end;
    private final int totalPage;

    public Pageable(int start, int limit, int size) {
        this.offset = Math.max(0, (start - 1) * limit);
        this.end = Math.min(offset + limit, size);
        this.totalPage = size % limit == 0? size / limit: size / limit + 1;

        if(size > 0 && offset > size){
            throw new PageSizeTooLargeException();
        }
    }

    public int getOffset() {
        return offset;
    }

    public int getEnd() {
        return end;
    }

    public int getTotalPage() {
        return totalPage;
    }
}

File: go/app/main/infrastructure/utils/pageable.go
package utils

import "mutualAttentionSystem/app/main/domain/errors"

type Pageable struct {
    Offset    int
    End       int
    TotalPage int
}

func NewPableable(start int, limit int, size int) (*Pageable, error) {
    offset := max(0, (start-1)*limit)
    end := min(offset+limit, size)
    totalPage := calculateTotalPage(limit, size)

    if size > 0 && offset > size {
        return nil, &errors.PageSizeTooLargeError{}
    }

    return &Pageable{offset, end, totalPage}, nil
}

func calculateTotalPage(limit int, size int) int {
    var totalPage int
    if size%limit == 0 {
        totalPage = size / limit
    } else {
        totalPage = size/limit + 1
    }
    return totalPage
}
